### DISCUSSION TITLE
how to test a function that contains a setTimeout?

### DISCUSSION THREAD

**Discussion by velly** [vitest-dev/vitest#2712] [Q&A]
I have an pinia action that contains setTimeout to call another action.
```
 actions:{
resetScheduler(): void {
    if (this.timeOutScheduler) {
      clearTimeout(this.timeOutScheduler);
    }
    if (this.tokenTimeOut) {
      let timeOutDelta = this.tokenTimeOut - Date.now();
      timeOutDelta = timeOutDelta > 0 ? timeOutDelta : 0;
      this.timeOutScheduler = setTimeout(this.timeoutToDoSomething, timeOutDelta);
    }
  },
 timeoutToDoSomething() {
  },
}
```
How can I test the action 'timeoutToDoSomething' was called once if this.tokenTimeOut is true.
Is there any method like waiting seconds?

```
it("if token is timout resetScheduler will work and timeoutToLogin would be called", async () => {
    const authStore = useAuth();
    const fakeAuth: AuthProtocal = JSON.parse(
      JSON.stringify(mockAuthObj)
    ) as AuthProtocal;
    fakeAuth.tokenTimeOutUTC = new Date(Date.now() - 1 * 1000).toISOString();
    authStore.loadAuthInfo(fakeAuth);
    const resetSchedulerSpy = vi.spyOn(authStore, "resetScheduler");
    const timeoutToLoginSpy = vi.spyOn(authStore, "timeoutToLogin");
    authStore.resetScheduler();
    await delay(2000);
    expect(resetSchedulerSpy).toHaveBeenCalledOnce();
    expect(timeoutToLoginSpy).toHaveBeenCalledOnce();
  });
```

**Comment by velly** [comment]
Now, I write a delay method and use await to workaroud:
`
function delay(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
`

`  it("if token is timout resetScheduler will work and timeoutToLogin would be called", async () => {
     ....
    await delay(2000);
    expect(authStore.timeoutToLogin).toHaveBeenCalledOnce();
  });`

**Reply by GRPdream** [reply]
Thanks bro, It works for me

**Reply by ty-lerscott** [reply]
correct me if I'm wrong but I also believe using `vi.useFakeTimers()` will resolve the timeout immediately so that you dont have to use that delay function
